precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+			.try_into()
+			.map_err(|_| revert("block number overflowed u64"))?;
+		let requested_block_number = blocks_after_current
+			.checked_add(current_block_number)
+			.ok_or(error("addition result overflowed u64"))?
+			.try_into()
+			.map_err(|_| revert("u64 addition result overflowed block number type"))?;
+		let request = Request {
+			refund_address,
+			contract_address,
+			fee,
+			gas_limit,
+			num_words,
+			salt,
+			info: RequestType::Local(requested_block_number),
+		};
+
+		let request_id: U256 = Pallet::<Runtime>::request_randomness(request)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?
+			.into();
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: EvmDataWriter::new().write(request_id).build(),
+		})
+	}
+	/// Fulfill a randomness request due to be fulfilled
+	fn fulfill_request(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let request_id = input.read::<U256>()?.low_u64();
+		// read all the inputs
+		let pallet_randomness::FulfillArgs {
+			request,
+			deposit,
+			randomness,
+		} = Pallet::<Runtime>::prepare_fulfillment(request_id)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?;
+		// check that randomness can be provided
+		ensure_can_provide_randomness::<Runtime>(
+			handle.code_address(),
+			request.gas_limit,
+			request.fee,
+			FULFILLMENT_OVERHEAD_ESTIMATED_COST,
+		)?;
+		// get gas before subcall
+		let before_remaining_gas = handle.remaining_gas();
+		provide_randomness(
+			handle,
+			request_id,
+			request.gas_limit,
+			request.contract_address.clone().into(),
+			randomness.into_iter().map(|x| H256(x)).collect(),
+		)?;
+		// get gas after subcall
+		let after_remaining_gas = handle.remaining_gas();
+		let gas_used: U256 = before_remaining_gas
+			.checked_sub(after_remaining_gas)
+			.ok_or(revert("Before remaining gas < After remaining gas"))?
+			.into();
+		// cost of execution is before_remaining_gas less after_remaining_gas
+		let cost_of_execution: BalanceOf<Runtime> = gas_used
+			.checked_mul(pallet_base_fee::Pallet::<Runtime>::base_fee_per_gas())
+			.ok_or(revert("Multiply gas used by base fee overflowed"))?
+			.try_into()
+			.map_err(|_| revert("amount is too large for provided balance type"))?;
+		// Finish fulfillment to
+		// refund cost of execution to caller
+		// refund excess fee to the refund_address
+		// remove request state
+		Pallet::<Runtime>::finish_fulfillment(
+			request_id,
+			request,
+			deposit,
+			&handle.context().caller,
+			cost_of_execution,
+		);
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: Default::default(),
+		})
+	}
+	/// Increase the fee used to refund fulfillment of the request
+	fn increase_request_fee(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let request_id = input.read::<U256>()?.low_u64();
```

Comment:
`low_u64()` has gotten us into trouble before. Should we ensure that all higher bits are unset? This might at the very least make it easier/safer for smart contract dev.

Reply:
Yes, I added that to quickly work on the test but @4meta5 we need a proper solution. Either checking it doesn't overflow or making the requestId u256